File: js/fib.js
// fib(n) returns nth number of fibonacci series
// n = 1 -> 1
// n = 0 -> 0


function fib(n) {
	const table = Array.from({length: n+1}, () => 0)
	table[1] = 1

	for (let i=0; i <= n; i++) {
		table[i+1] += table[i]
		table[i+2] += table[i]
	}
	return table[n]
}

console.log(fib(6))

File: cpp/fib.cpp
// Find fibonacci number at n position
// fib(3) -> 1

#include <iostream>
#include <map>
#include <functional>

using namespace std;

long fib(int n) {

	function<long(int)> impl = [memo=move(map<long, long>()), &impl] (int n) mutable -> long {
		if (n == 0) return 0;
		if (n == 1 || n == 2) return 1;
		map<long, long>::iterator l;
		if ( (l = memo.find(n)) != memo.end()) {
			return (*l).second;
		}

		long m = impl(n-1) + impl(n-2);

		memo.emplace(n, m);
		return m;
	};

	return impl(n);
}

int main() {
	cout << fib(100) << endl;
}
